Fix stale lock issue with ONE_PHONE_NUMBER_MULTIPLE_CONTACTS flag

dannyroberts opened the pull request:
https://dimagi-dev.atlassian.net/browse/SAAS-11045

##### SUMMARY
It appears that Keyword SMS Surveys that have a single label and no other questions leave the form session with `session.session_is_open = False` somehow without actually calling `session.close()` (which also releases the channel claim), and so the channel claim is left around. Currently this prevents other things from happening until the channel claim is explicitly cleared with `clear_stale_channel_claim` at a later time, but that call is put in a few places somewhat arbitrarily and I hadn't expected to rely on it. This change makes it call the same log `clear_stale_channel_claim` does every time you claim a channel, thus treating a channel claimed by a stale session the same as a channel not claimed for most purposes.

##### FEATURE FLAG
ONE_PHONE_NUMBER_MULTIPLE_CONTACTS

dannyroberts committed:
Always clear stale channel claim when claiming

dannyroberts committed:
Add test for auto-clearing stale sessions from channel on claim

dannyroberts committed:
Make channel_is_available_for_session auto-clear stale sessions as well

dannyroberts committed:
Make clear_stale_channel_claim implement sticky contact behavior as well

Previously, release_channel_for_session would leave the contact in
in the running session info (even while setting the session_id to None)
so that incoming texts would be preferentially treated as coming from
the last active contact; but clear_stale_channel_claim would delete both.

Now, clear_stale_channel_claim adopts this same sticky-contact behavior
as well.

millerdev commented on corehq/apps/smsforms/models.py:
```diff
@@ -296,7 +296,7 @@ def claim_channel_for_session(cls, session):
         channel = session.get_channel()
         with cls._critical_section(channel):
-            if cls.channel_is_available_for_session(session):
+            if cls._channel_is_available_for_session(session) or cls._clear_stale_channel_claim(channel):
```
Looks like `_channel_is_available_for_session` calls `_clear_stale_channel_claim`. Why call it again on this line?

dannyroberts replied:
Thanks for pointing that out, you're right that was an oversight. I added this line before realizing I should move it down into `_channel_is_available_for_session`, and ended up leaving it in both. (The blackbox behavior is still the same so tests didn't fail, but you're right that it's entirely redundant.)

proteusvacuum commented on corehq/apps/smsforms/tests/test_session_synchronization.py:
```diff
@@ -47,6 +47,34 @@ def test_session_synchronization():
+    # Claim for the channel now succeeds
+    assert XFormsSessionSynchronization.channel_is_available_for_session(session_a_2)
+    assert XFormsSessionSynchronization.claim_channel_for_session(session_a_2)
+    # And it is now the active session for the channel
```
:tada:

dannyroberts committed:
Remove redundant call to cls._clear_stale_channel_claim

This call has been moved into _channel_is_available_for_session making it redundant
as Daniel pointed out in
https://github.com/dimagi/commcare-hq/pull/27976/files#r446939836

dannyroberts committed:
Improve testing for auto-clearing stale channel claim

dannyroberts committed:
Remove redundant call to clear_stale_channel_claim

handle_survey_actions no longer needs to call clear_stale_channel_claim directly
because channel_is_available_for_session, which handle_survey_actions calls,
now calls it

dannyroberts commented:
@millerdev and @proteusvacuum thanks for the review. Your comments inspired a couple more changes, but I think once those are reviewed I'm happy with the state of this change set and would be ready to release it today
